core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go

```diff
@@ -227,15 +227,20 @@ func (bs *BlockSubscriber) processHead(h *evmtypes.Head) {
 	// head parent is a linked list with EVM finality depth
 	// when re-org happens, new heads will have pointers to the new blocks
 	i := int64(0)
-	for cp := h; cp != nil; cp = cp.Parent {
-		if cp != h && bs.blocks[cp.Number] != cp.Hash.Hex() {
-			bs.lggr.Warnf("overriding block %d old hash %s with new hash %s due to re-org", cp.Number, bs.blocks[cp.Number], cp.Hash.Hex())
+	cp := new(evmtypes.Head)
+	*cp = *h
+	for ; ; cp = cp.Parent {
+		if cp.Number != h.Number && bs.blocks[cp.Number] != cp.Hash.Hex() {
+			bs.lggr.Warnf("h.Number=%d cp.Number=%d overriding block %d old hash %s with new hash %s due to re-org or missing previous heads", h.Number, cp.Number, cp.Number, bs.blocks[cp.Number], cp.Hash.Hex())
 		}
 		bs.blocks[cp.Number] = cp.Hash.Hex()
 		i++
 		if i > bs.blockSize {
 			break
 		}
+		if cp.Parent == nil || bs.blocks[cp.Parent.Number] == cp.Parent.BlockHash().Hex() {
+			break
+		}
```

Comment:
i think to simplify you can just check `bs.blocks[cp.Number] != cp.Hash.Hex()` at start of loop and exit there, instead of checking for nil parent and separate check here

Reply:
done